From fb403ec096197c689cf2eb3afc703fd64defbac1 Mon Sep 17 00:00:00 2001
From: Bichao Zheng <bichao.zheng@amlogic.com>
Date: Thu, 20 Apr 2023 16:23:24 +0800
Subject: [PATCH] nand: fix nand bugs [1/1]

PD#SWPL-120491
PD#SWPL-129754
PD#SWPL-155674
PD#SWPL-249674
PD#TV-210454

Problem:
1.wrong select onfi mode
2.erase flash error during OTA upgrade
3.strncmp can't match full id 01h,DCh,00h,05h,04h(S34ML04G300BHI00)
and 01h,DCh,00h,1Ah,00h(S34ML04G300TFI000)
4.add GD9FU4G8F4D support
5.I/O error after resume

Solution:
1.get onfi mode from id table
2.change the check way of badblock
3.use memcmp to compare ID memory instead strncmp
4.timing support max tREA 9ns
5.reset chip after resume

Verify:
1-2:s4

Change-Id: Ie747ceb95f1045c2acaaf2afb6862c9b05d36d8b
Signed-off-by: Bichao Zheng <bichao.zheng@amlogic.com>
Signed-off-by: zhikui.cui <zhikui.cui@amlogic.com>
---
 drivers/mtd/nand/raw/nand_base.c    | 15 ++++++++-
 drivers/mtd/nand/raw/nand_bbt.c     | 11 ++++++
 drivers/mtd/nand/raw/nand_timings.c | 52 +++++++++++++++++++++++++++++
 include/linux/mtd/rawnand.h         | 17 ++++++++++
 4 files changed, 94 insertions(+), 1 deletion(-)

diff --git a/drivers/mtd/nand/raw/nand_base.c b/drivers/mtd/nand/raw/nand_base.c
index 3e1844bfb808..3fd40973c88b 100644
--- a/drivers/mtd/nand/raw/nand_base.c
+++ b/drivers/mtd/nand/raw/nand_base.c
@@ -967,6 +967,10 @@ int nand_choose_best_sdr_timings(struct nand_chip *chip,
 	return ret;
 }
 
+#if IS_ENABLED(CONFIG_AMLOGIC_MTD_NAND)
+EXPORT_SYMBOL_GPL(nand_choose_best_sdr_timings);
+#endif
+
 /**
  * nand_choose_best_nvddr_timings - Pick up the best NVDDR timings that both the
  *                                  NAND controller and the NAND chip support
@@ -4835,6 +4839,9 @@ static void nand_resume(struct mtd_info *mtd)
 {
 	struct nand_chip *chip = mtd_to_nand(mtd);
 
+#if IS_ENABLED(CONFIG_AMLOGIC_MTD_NAND)
+	nand_reset(chip, 0);
+#endif
 	mutex_lock(&chip->lock);
 	if (chip->suspended) {
 		if (chip->ops.resume)
@@ -5086,7 +5093,11 @@ static bool find_full_id_nand(struct nand_chip *chip,
 
 	memorg = nanddev_get_memorg(&chip->base);
 
+#if IS_ENABLED(CONFIG_AMLOGIC_MTD_NAND)
+	if (!memcmp(type->id, id_data, type->id_len)) {
+#else
 	if (!strncmp(type->id, id_data, type->id_len)) {
+#endif
 		memorg->pagesize = type->pagesize;
 		mtd->writesize = memorg->pagesize;
 		memorg->pages_per_eraseblock = type->erasesize /
@@ -5409,7 +5420,9 @@ static int nand_detect(struct nand_chip *chip, struct nand_flash_dev *type)
 		chip->options |= NAND_ROW_ADDR_3;
 
 	chip->badblockbits = 8;
-
+#if IS_ENABLED(CONFIG_AMLOGIC_MTD_NAND)
+	chip->type = type;
+#endif
 	nand_legacy_adjust_cmdfunc(chip);
 
 	pr_info("device found, Manufacturer ID: 0x%02x, Chip ID: 0x%02x\n",
diff --git a/drivers/mtd/nand/raw/nand_bbt.c b/drivers/mtd/nand/raw/nand_bbt.c
index a8fba5f39f59..9dfcef0d846e 100644
--- a/drivers/mtd/nand/raw/nand_bbt.c
+++ b/drivers/mtd/nand/raw/nand_bbt.c
@@ -75,9 +75,16 @@
 
 static inline uint8_t bbt_get_entry(struct nand_chip *chip, int block)
 {
+#if !IS_ENABLED(CONFIG_AMLOGIC_MTD_NAND)
 	uint8_t entry = chip->bbt[block >> BBT_ENTRY_SHIFT];
 	entry >>= (block & BBT_ENTRY_MASK) * 2;
 	return entry & BBT_ENTRY_MASK;
+#else
+	if (chip->bbt)
+		return	chip->bbt[block];
+
+	return 1;
+#endif
 }
 
 static inline void bbt_mark_entry(struct nand_chip *chip, int block,
@@ -1457,6 +1464,7 @@ int nand_isbad_bbt(struct nand_chip *this, loff_t offs, int allowbbt)
 	if (mtd_check_expert_analysis_mode())
 		return 0;
 
+#if !IS_ENABLED(CONFIG_AMLOGIC_MTD_NAND)
 	switch (res) {
 	case BBT_BLOCK_GOOD:
 		return 0;
@@ -1466,6 +1474,9 @@ int nand_isbad_bbt(struct nand_chip *this, loff_t offs, int allowbbt)
 		return allowbbt ? 0 : 1;
 	}
 	return 1;
+#else
+	return res ? 1 : 0;
+#endif
 }
 
 /**
diff --git a/drivers/mtd/nand/raw/nand_timings.c b/drivers/mtd/nand/raw/nand_timings.c
index 7b41afc372d2..033917df9917 100644
--- a/drivers/mtd/nand/raw/nand_timings.c
+++ b/drivers/mtd/nand/raw/nand_timings.c
@@ -290,6 +290,54 @@ static const struct nand_interface_config onfi_sdr_timings[] = {
 			.tWW_min = 100000,
 		},
 	},
+#ifdef CONFIG_AMLOGIC_NAND
+	/**
+	 * Mode 6 for GD9FU4G8F4D
+	 * GD9FU4G8F4D tREA is 9ns, which does not comply with
+	 * ONFI TIMING MODE 5
+	 */
+	{
+		.type = NAND_SDR_IFACE,
+		.timings.sdr = {
+			.tCCS_min = 500000,
+			.tR_max = 200000000,
+			.tADL_min = 400000,
+			.tALH_min = 5000,
+			.tALS_min = 10000,
+			.tAR_min = 10000,
+			.tCEA_max = 25000,
+			.tCEH_min = 20000,
+			.tCH_min = 5000,
+			.tCHZ_max = 30000,
+			.tCLH_min = 5000,
+			.tCLR_min = 10000,
+			.tCLS_min = 10000,
+			.tCOH_min = 15000,
+			.tCS_min = 15000,
+			.tDH_min = 5000,
+			.tDS_min = 7000,
+			.tFEAT_max = 1000000,
+			.tIR_min = 0,
+			.tITC_max = 1000000,
+			.tRC_min = 20000,
+			.tREA_max = 9000,
+			.tREH_min = 7000,
+			.tRHOH_min = 15000,
+			.tRHW_min = 100000,
+			.tRHZ_max = 100000,
+			.tRLOH_min = 5000,
+			.tRP_min = 10000,
+			.tRR_min = 20000,
+			.tRST_max = 500000000,
+			.tWB_max = 100000,
+			.tWC_min = 20000,
+			.tWH_min = 7000,
+			.tWHR_min = 80000,
+			.tWP_min = 10000,
+			.tWW_min = 100000,
+		},
+	},
+#endif
 };
 
 static const struct nand_interface_config onfi_nvddr_timings[] = {
@@ -735,3 +783,7 @@ void onfi_fill_interface_config(struct nand_chip *chip,
 	else
 		return onfi_fill_nvddr_interface_config(chip, iface, timing_mode);
 }
+
+#if IS_ENABLED(CONFIG_AMLOGIC_MTD_NAND)
+EXPORT_SYMBOL_GPL(onfi_fill_interface_config);
+#endif
diff --git a/include/linux/mtd/rawnand.h b/include/linux/mtd/rawnand.h
index e84522e31301..2dcbb18de6b5 100644
--- a/include/linux/mtd/rawnand.h
+++ b/include/linux/mtd/rawnand.h
@@ -1342,6 +1342,9 @@ struct nand_chip {
 	/* Externals */
 	struct nand_controller *controller;
 	struct nand_ecc_ctrl ecc;
+#if IS_ENABLED(CONFIG_AMLOGIC_MTD_NAND)
+	struct nand_flash_dev *type;
+#endif
 	void *priv;
 };
 
@@ -1470,6 +1473,9 @@ struct nand_flash_dev {
 		uint16_t strength_ds;
 		uint16_t step_ds;
 	} ecc;
+#if IS_ENABLED(CONFIG_AMLOGIC_MTD_NAND)
+	int onfi_timing_mode_default;
+#endif
 };
 
 int nand_create_bbt(struct nand_chip *chip);
@@ -1587,6 +1593,17 @@ int nand_read_page_hwecc_oob_first(struct nand_chip *chip, uint8_t *buf,
 int nand_scan_with_ids(struct nand_chip *chip, unsigned int max_chips,
 		       struct nand_flash_dev *ids);
 
+#if IS_ENABLED(CONFIG_AMLOGIC_MTD_NAND)
+void onfi_fill_interface_config(struct nand_chip *chip,
+				struct nand_interface_config *iface,
+				enum nand_interface_type type,
+				unsigned int timing_mode);
+
+int nand_choose_best_sdr_timings(struct nand_chip *chip,
+				 struct nand_interface_config *iface,
+				 struct nand_sdr_timings *spec_timings);
+#endif
+
 static inline int nand_scan(struct nand_chip *chip, unsigned int max_chips)
 {
 	return nand_scan_with_ids(chip, max_chips, NULL);
-- 
2.47.1

